dGame/dGameMessages/GameMessages.cpp

```diff
@@ -6342,36 +6342,57 @@ void GameMessages::SendUpdateInventoryUi(LWOOBJID objectId, const SystemAddress&
 	SEND_PACKET;
 }
 
-void GameMessages::DisplayTooltip::Send() const {
-	CBITSTREAM;
-	CMSGHEADER;
+namespace GameMessages {
+	void GameMsg::Send(const SystemAddress& sysAddr) const {
+		CBITSTREAM;
+		CMSGHEADER;
 
-	bitStream.Write(target);
-	bitStream.Write(msgId);
-
-	bitStream.Write(doOrDie);
-	bitStream.Write(noRepeat);
-	bitStream.Write(noRevive);
-	bitStream.Write(isPropertyTooltip);
-	bitStream.Write(show);
-	bitStream.Write(translate);
-	bitStream.Write(time);
-	bitStream.Write<int32_t>(id.size());
-	bitStream.Write(id);
-
-	std::string toWrite;
-	for (const auto* item : localizeParams) {
-		toWrite += item->GetString() + "\n";
-	}
-	if (!toWrite.empty()) toWrite.pop_back();
-	bitStream.Write<int32_t>(toWrite.size());
-	bitStream.Write(GeneralUtils::ASCIIToUTF16(toWrite));
-	if (!toWrite.empty()) bitStream.Write<uint16_t>(0x00); // Null Terminator
-
-	bitStream.Write<int32_t>(imageName.size());
-	bitStream.Write(imageName);
-	bitStream.Write<int32_t>(text.size());
-	bitStream.Write(text);
+		bitStream.Write(target); // Who this message will be sent to on the (a) client
+		bitStream.Write(msgId); // the ID of this message
 
-	SEND_PACKET;
+		Serialize(bitStream); // write the message data
+
+		// Send to everyone if someone sent unassigned system address, or to one specific client.
+		if (sysAddr == UNASSIGNED_SYSTEM_ADDRESS) {
+			SEND_PACKET_BROADCAST;
+		} else {
+			SEND_PACKET;
+		}
+	}
+
+	void DisplayTooltip::Serialize(RakNet::BitStream& bitStream) const {
+		bitStream.Write(doOrDie);
+		bitStream.Write(noRepeat);
+		bitStream.Write(noRevive);
+		bitStream.Write(isPropertyTooltip);
+		bitStream.Write(show);
+		bitStream.Write(translate);
+		bitStream.Write(time);
+		bitStream.Write<int32_t>(id.size());
+		bitStream.Write(id);
+
+		std::string toWrite;
+		for (const auto* item : localizeParams) {
+			toWrite += item->GetString() + "\n";
+		}
+		if (!toWrite.empty()) toWrite.pop_back();
+		bitStream.Write<int32_t>(toWrite.size());
+		bitStream.Write(GeneralUtils::ASCIIToUTF16(toWrite));
+		if (!toWrite.empty()) bitStream.Write<uint16_t>(0x00); // Null Terminator
+
+		bitStream.Write<int32_t>(imageName.size());
+		bitStream.Write(imageName);
+		bitStream.Write<int32_t>(text.size());
+		bitStream.Write(text);
+	}
+
+	void UseItemOnClient::Serialize(RakNet::BitStream& bitStream) const {
+		bitStream.Write(itemLOT);
+		bitStream.Write(itemToUse);
+		bitStream.Write(itemType);
+		bitStream.Write(playerId);
+		bitStream.Write(targetPosition.x);
+		bitStream.Write(targetPosition.y);
+		bitStream.Write(targetPosition.z);
```

Comment:
Since you asked for some feedback on the design: Doesn't Raknet provide a way to serialize all the fields of a struct using one function call?

Reply:
it does allow this, however this relies on the structure never changing and we need it to match something in not our code, hence why I personally manually write each field of a struct (you can just call .Write(targetPosition) however this will include packing bytes, which we dont have and since we cant control the structure well, i'd rather not risk it and set a pattern that may be broken in some contexts).  Do you still think its worth it to write the position struct as is?

Reply:
an alternative to make sure its correct while allowing the direct write of a struct would be to override this template in our bitstream utils like we do with string.  I could do that here if its wanted

Reply:
What about a library to automatically deserialize the struct without packing bytes? Maybe something like this? https://github.com/p-ranav/alpaca

It would be so very nice to just pass a struct to a function and have it automatically be written as a bitstream.

Reply:
My understanding is the "behind the scenes" there is you're basically just using structured unbinding and template specialization to generate the appropriate function calls for each field

Reply:
Yes it would be nice, however alpaca supports from a glance _byte_ arrays, not bit arrays, which is what we would need here.